File: clients/php/src/lib/Kafka/Message.php
<?php

/**
 * A message. The format of an N byte message is the following:
 * 1 byte "magic" identifier to allow format changes
 * 4 byte CRC32 of the payload
 * N - 5 byte payload
 *
 * @author Lorenzo Alberton <l.alberton@quipo.it>
 */
class Kafka_Message
{
	/*
	private $currentMagicValue = Kafka_Encoder::CURRENT_MAGIC_VALUE;
	private $magicOffset   = 0;
	private $magicLength   = 1;
	private $crcOffset     = 1; // MagicOffset + MagicLength
	private $crcLength     = 4;
	private $payloadOffset = 5; // CrcOffset + CrcLength
	private $headerSize    = 5; // PayloadOffset
	*/
	
	private $payload = null;
	private $size    = 0;
	private $crc     = false;
	
	/**
	 * @param string $data
	 */
	public function __construct($data) {
		$this->payload = substr($data, 5);
		$this->crc     = crc32($this->payload);
		$this->size    = strlen($this->payload);
	}
	
	/**
	 * @return string
	 */
	public function encode() {
		return Kafka_Encoder::encode_message($this->payload);
	}
	
	/**
	 * @return integer
	 */
	public function size() {
		return $this->size;
	}
  
	/**
	 * @return integer
	 */
	public function magic() {
		return Kafka_Encoder::CURRENT_MAGIC_VALUE;
	}
	
	/**
	 * @return integer
	 */
	public function checksum() {
		return $this->crc;
	}
	
	/**
	 * @return string
	 */
	public function payload() {
		return $this->payload;
	}
	
	/**
	 * @return boolean
	 */
	public function isValid() {
		return ($this->crc === crc32($this->payload));
	}
  
	/**
	 * @return string
	 */
	public function __toString() {
		return 'message(magic = ' . Kafka_Encoder::CURRENT_MAGIC_VALUE . ', crc = ' . $this->crc .
			', payload = ' . $this->payload . ')';
	}
}

File: core/src/main/scala/kafka/message/Message.scala
/*
 * Copyright 2010 LinkedIn
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.message

import java.nio._
import java.nio.channels._
import java.util.zip.CRC32
import java.util.UUID
import kafka.utils._

/**
 * Message byte offsets
 */
object Message {
  val CurrentMagicValue: Byte = 1
  val MagicOffset = 0
  val MagicLength = 1
  val AttributeOffset = MagicOffset + MagicLength
  val AttributeLength = 1
  
  def CrcOffset(magic:Byte): Int = magic match {
    case 0 => MagicOffset + MagicLength
    case _ => AttributeOffset + AttributeLength
  }
  
  val CrcLength = 4
  def PayloadOffset(magic:Byte): Int = CrcOffset(magic) + CrcLength
  def HeaderSize(magic:Byte): Int = PayloadOffset(magic)
  val MinHeaderSize = HeaderSize(0);
  // Attribute masks follow
  val COMPRESSION_CODEC_MASK:Int = 0x03  // 2 bits to hold the compression codec. 0 is reserved to indicate no compression
  
  
  // Defaults follow
  val NO_COMPRESSION:Int = 0
}

/**
 * A message. The format of an N byte message is the following:
 * 1 byte "magic" identifier to allow format changes
 * 1 byte "attributes" identifier to allow annotations on the message independent of the version (e.g. compression enabled, type of codec used)
 * 4 byte CRC32 of the payload
 * N - 8 byte payload
 * 
 */
class Message(val buffer: ByteBuffer) {
  
  import kafka.message.Message._
    
  
  private def this(checksum: Long, bytes: Array[Byte], compressionCodec:Int) = {
    this(ByteBuffer.allocate(Message.HeaderSize(Message.CurrentMagicValue) + bytes.length))
    buffer.put(CurrentMagicValue)
    var attributes:Byte = 0
    if (compressionCodec >0) {
      attributes =  (attributes | (Message.COMPRESSION_CODEC_MASK & compressionCodec)).toByte     
    }
    buffer.put(attributes)
    Utils.putUnsignedInt(buffer, checksum)
    buffer.put(bytes)
    buffer.rewind()
  }
  def this(checksum:Long, bytes:Array[Byte]) = this(checksum, bytes, Message.NO_COMPRESSION)
  
  def this(bytes: Array[Byte], compressionCodec:Int) = {
    //Note: we're not crc-ing the attributes header, so we're susceptible to bit-flipping there
    this(Utils.crc32(bytes), bytes, compressionCodec)
  }
  def this(bytes:Array[Byte]) = this(bytes, Message.NO_COMPRESSION)
  
  def size: Int = buffer.limit
  
  def payloadSize: Int = size - HeaderSize(magic)
  
  def magic: Byte = buffer.get(MagicOffset)
  
  def attributes: Byte = buffer.get(AttributeOffset)
  
  def compressionCodec:Int = {
    buffer.get(AttributeOffset) & COMPRESSION_CODEC_MASK;
  }
  
  def isCompressed:Boolean = magic match {
    case 0 => false
    case _ => compressionCodec > 0
  }
  
  def checksum: Long = Utils.getUnsignedInt(buffer, CrcOffset(magic))
  
  def payload: ByteBuffer = {
    var payload = buffer.duplicate
    payload.position(HeaderSize(magic))
    payload = payload.slice()
    payload.limit(payloadSize)
    payload.rewind()
    payload
  }
  
  def isValid: Boolean = 
    checksum == Utils.crc32(buffer.array, buffer.position + buffer.arrayOffset + PayloadOffset(magic), payloadSize)
     
  def serializedSize: Int = 4 /* int size*/ + buffer.limit
   
  def serializeTo(serBuffer:ByteBuffer) = {
    serBuffer.putInt(buffer.limit)
    serBuffer.put(buffer.duplicate)
  }

  override def toString(): String = 
    "message(magic = " + magic + ", attributes = " + attributes + ", crc = " + checksum + 
    ", payload = " + payload + ")"
  
  override def equals(any: Any): Boolean = {
    any match {
      case that: Message => size == that.size && attributes == that.attributes && checksum == that.checksum && payload == that.payload && magic == that.magic
      case _ => false
    }
  }
  
  override def hashCode(): Int = buffer.hashCode
  
}
